### PR TITLE
Post scans api

### PR DESCRIPTION
**Pull request by gdetrez**
Adds the following endpoints:
- `GET /scans`: return metadata for all scans in the database
- `POST /scans`: add a scan
- `GET /scans/ID`: get metadata for a particular scan
- `GET /scans/ID/image`: get the image for a particular scan

The images are saved in the project directory (e.g. `/somprojects`) but the name of the image is not exactly the same as before: Instead of `{project}_{index}_{timedelta}.tiff` it's now `{project}_{timestamp}.tiff`.  If the compilation/analysis only requires the images to be sorted it should be fine. Otherwise with a bit more work we can probably update the code to estimate both the index and the timedelta from the job's start time.

### CODE REVIEW

File: scanomatic/io/imagestore.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from datetime import datetime
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/io/imagestore.py]: fixed

File: tests/integration/api/conftest.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import json
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/integration/api/conftest.py]: fixed

File: tests/unit/io/test_imagestore.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+from datetime import datetime
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/unit/io/test_imagestore.py]: fixed

File: tests/unit/io/test_scanning_store.py [hunk 4]
```diff
@@ -7,9 +7,11 @@
 from pytz import utc
 
 from scanomatic.io.scanning_store import (
-    ScanningStore, Scanner, ScanJobUnknownError, ScanJobCollisionError,
+    ScanningStore, ScanJobCollisionError, ScanJobUnknownError, Scanner,
+    DuplicateIdError, UnknownIdError
 )
 from scanomatic.models.scanjob import ScanJob
+from scanomatic.models.scan import Scan
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/unit/io/test_scanning_store.py]: fixed

File: scanomatic/ui_server/ui_server.py [hunk 5]
```diff
@@ -54,7 +55,7 @@ def launch_server(host, port, debug):
     global _URL, _DEBUG_MODE
     _DEBUG_MODE = debug
     app = Flask("Scan-o-Matic UI", template_folder=Paths().ui_templates)
-    app.config['scanstore'] = ScanStore(Config().paths.projects_root)
+    app.config['imagestore'] = ImageStore(Config().paths.projects_root)
```
Comment (local-minimum) [inline scanomatic/ui_server/ui_server.py]: There's an `add_config` function on line 127. Either move this there or that here.

File: tests/integration/api/conftest.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+
+@pytest.fixture
+def app(tmpdir):
+    app = Flask(__name__)
+    app.register_blueprint(scanners_api.blueprint, url_prefix="/scanners")
+    app.register_blueprint(scan_jobs_api.blueprint, url_prefix="/scan-jobs")
+    app.register_blueprint(scans_api.blueprint, url_prefix="/scans")
+    add_configs(app)
+    app.config['imagestore'] = ImageStore(str(tmpdir))
```
Comment (local-minimum) [inline tests/integration/api/conftest.py]: leaning towards that this should be part of `add_configs`

**Comment by local-minimum** [comment]
I think the new file names are fine. If anything it would be the job of the compilation-step to figure out the indices and skipping indices when images are missing.

File: scanomatic/ui_server/scans_api.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            scanjob_id=args.scanjob_id,
+        )
+        # ⚠ Without support for transactions, the order of the following calls
+        # matter: we add the metadata to the database only if the image has
+        # been stored on the disk. That way, the worst that can happen is
+        # an orphan image file. Otherwise, the database may thing it has
+        # the scan whereas the image is not there. (I.e. the source of truth
+        # is the database).
+        imagestore.put(args.image.read(), scan)
+        db.add_scan(scan)
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/ui_server/scans_api.py]: I don't think so: in the (very unlikely) event that the generated id already exist, I think we actually want the scanner to try uploading the same scan again.

File: scanomatic/io/scanning_store.py [hunk 8]
```diff
@@ -98,3 +107,20 @@ def get_current_scanjob(self, scanner_id, timepoint):
...
     def has_current_scanjob(self, scanner_id, timepoint):
         return self.get_current_scanjob(scanner_id, timepoint) is not None
+
+    def add_scan(self, scan):
+        if scan.scanjob_id not in self._scanjobs:
+            raise ScanJobUnknownError
+        if scan.id in self._scans:
+            raise DuplicateIdError
+        self._scans[scan.id] = scan
+
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/io/scanning_store.py]: yagni

File: scanomatic/ui_server/ui_server.py (hunk 5 above)
Comment (gdetrez) [reply scanomatic/ui_server/ui_server.py]: I know but that creates problems:
- I use a different path in integration tests
- `add_config` is used in some unit tests. That's not ideal in itself so I'd rather not add to it.

File: scanomatic/ui_server/scans_api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (local-minimum) [reply scanomatic/ui_server/scans_api.py]: Don't we run the risk of it getting stuck on an image?

File: scanomatic/ui_server/ui_server.py (hunk 5 above)
Comment (local-minimum) [reply scanomatic/ui_server/ui_server.py]: How about removing it then?

**Pull request merged**